src/validations/html/rules/MismatchedTag.js

```diff
@@ -0,0 +1,139 @@
+import {List, Map, Record, Set} from 'immutable';
+import Code from '../rules/Code';
+
+const Location = Record({row: 0, column: 0});  // eslint-disable-line new-cap
+
+export default class {
```

Comment:
@tenyoung795 I definitely think the `misplaced-close-tag` error is an indispensable part of the functionality in this PR, but I think my proposed structure handles the example you give here, specifically:

> If the closing tag we have encountered matches an entry in collection 1, remove that entry from collection 1, and add the open/close pair to collection 2

Am I thinking about it wrong?